src/data/polynomial/mirror.lean

```diff
@@ -183,4 +184,124 @@ end
 
 end mirror
 
+section norm2
+
+variables {R : Type*} [comm_ring R] (p q : polynomial R)
+
+noncomputable def norm2 : quadratic_form R (polynomial R) :=
+finsupp.norm_sq
+
+lemma norm2_apply : p.norm2 = p.sum (λ k c, c ^ 2) :=
+finsupp.norm_sq_apply p
+
+lemma norm2_eq_sum_of_support {s : finset ℕ}
+  (h : p.support ⊆ s) : p.norm2 = s.sum (λ k, (p.coeff k) ^ 2) :=
+p.norm2_apply.trans (finset.sum_subset h (λ k h1 h2, by
+{ rw [finsupp.not_mem_support_iff.mp h2], exact zero_pow zero_lt_two }))
+
+lemma norm2_monomial (k : ℕ) (a : R) : (monomial k a).norm2 = a ^ 2 :=
+by rw [norm2_eq_sum_of_support _ (support_monomial' k a),
+  finset.sum_singleton, coeff_monomial, if_pos rfl]
+
+lemma norm2_C (a : R) : (C a).norm2 = a ^ 2 := norm2_monomial 0 a
+
+lemma norm2_zero : (0 : polynomial R).norm2 = 0 := by rw [←C_0, norm2_C, pow_two, zero_mul]
+
+lemma norm2_eq_mul_reverse_coeff :
+  p.norm2 = (p * p.mirror).coeff (p.nat_degree + p.nat_trailing_degree) :=
+begin
+  have h : p.support ⊆ finset.range (p.nat_degree + p.nat_trailing_degree).succ :=
+  λ x hx, finset.mem_range_succ_iff.mpr ((le_nat_degree_of_ne_zero
+    (mem_support_iff_coeff_ne_zero.mp hx)).trans (nat.le_add_right _ _)),
+  rw [eq_comm, p.norm2_eq_sum_of_support h, coeff_mul,
+      finset.nat.sum_antidiagonal_eq_sum_range_succ_mk],
+  apply finset.sum_congr rfl,
+  intros k hk,
+  rw [pow_two, coeff_mirror, ←rev_at_le (finset.mem_range_succ_iff.mp hk), rev_at_invol],
+end
+
+-- `p.nat_degree` can be recovered from `p * p.mirror`
+lemma nat_degree_mul_mirror {R : Type*} [integral_domain R] (p : polynomial R) :
+  (p * p.mirror).nat_degree = 2 * p.nat_degree :=
+begin
+  by_cases hp : p = 0,
+  { rw [hp, zero_mul, nat_degree_zero, mul_zero] },
+  { rw [nat_degree_mul hp (mt p.mirror_eq_zero.mp hp), mirror_nat_degree, two_mul] },
+end
+
+-- `p.nat_trailing_degree` can be recovered from `p * p.mirror`
+lemma nat_trailing_degree_mul_mirror {R : Type*} [integral_domain R] (p : polynomial R) :
+  (p * p.mirror).nat_trailing_degree = 2 * p.nat_trailing_degree :=
+begin
+  by_cases hp : p = 0,
+  { rw [hp, zero_mul, nat_trailing_degree_zero, mul_zero] },
+  { rw [nat_trailing_degree_mul hp (mt p.mirror_eq_zero.mp hp),
+        mirror_nat_trailing_degree, two_mul] },
+end
+
+-- `p.norm2` can be recovered from `p * p.mirror`
+lemma central_coeff_mul_mirror {R : Type*} [integral_domain R] (p : polynomial R) :
+  (p * p.mirror).coeff (((p * p.mirror).nat_degree +
+    (p * p.mirror).nat_trailing_degree) / 2) = p.norm2 :=
+by rw [nat_degree_mul_mirror, nat_trailing_degree_mul_mirror,  ←mul_add,
+  nat.mul_div_cancel_left _ zero_lt_two, norm2_eq_mul_reverse_coeff]
+
+lemma norm2_nonneg {R : Type*} [linear_ordered_comm_ring R] (p : polynomial R) : 0 ≤ norm2 p :=
```

Comment:
Can you state this and all the following lemmas in the context of `finsupp` and put them in `quadratic_form`?